core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -92,7 +92,13 @@ class ZooKeeperClient(connectString: String,
   metricNames += "SessionState"
 
   expiryScheduler.startup()
-  waitUntilConnected(connectionTimeoutMs, TimeUnit.MILLISECONDS)
+  try {
+    waitUntilConnected(connectionTimeoutMs, TimeUnit.MILLISECONDS)
+  } catch {
+    case e: ZooKeeperClientTimeoutException =>
+      close()
+      throw e
+  }
 
```

Comment:
Thanks for updating, I still think we should use a finally block instead of catching a specific exception.

Reply:
Oh wait. I was wrong, we can't use finally here. But we should probably catch a more general exception?

Reply:
Yeah, we can user ZooKeeperClientException, which parent is parent class for zk excpetions

Reply:
Right, as other exceptions can be thrown too. But to be honest, we might as well catch `Throwable` since we rethrow he exception anyway. We just wan to avoid leaking the resource.

Reply:
Updated the PR with ``` Throwable``` .   Not sure, if we can catch parent constructor  exception in subclass constructor. For now, added a test based on zk thread count.